[BEAM-1301] Support for BigQuery table description

ravwojdyla opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

dhalperi commented:
R: @dhalperi

Sorry, did not see this. Thanks Rav! Will take a look tomorrow.

dhalperi commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:
```diff
@@ -1716,21 +1723,26 @@ public Bound() {
             null /* jsonSchema */,
             CreateDisposition.CREATE_IF_NEEDED,
             WriteDisposition.WRITE_EMPTY,
+            null /*tableDescription */,
```
nit: `/* table` (with space)

ravwojdyla replied:
Fixed.

dhalperi commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:
```diff
@@ -2129,7 +2153,10 @@ public void populateDisplayData(DisplayData.Builder builder) {
+              .withLabel("Validation Enabled"), true)
+            .addIfNotNull(DisplayData.item("tableDescription", tableDescription)
+              .withLabel("Table description"));
+
```
drop blank line?

ravwojdyla replied:
Fixed.

dhalperi commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:
```diff
@@ -2374,6 +2408,13 @@ private void load(
           switch (jobStatus) {
             case SUCCEEDED:
+              if (tableDescription != null) {
+                datasetService.patchTableDescription(
```
Note to self and reviewers: confirmed that you can't set the table description from the load job itself: https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.load

dhalperi commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:
```diff
@@ -2374,6 +2408,13 @@ private void load(
+                    ref.getDatasetId(),
+                    ref.getTableId(),
+                    tableDescription);
+              }
```
you probably want this to run after copy job too, because for a very large table we load to temp tables then copy to the final table.

ravwojdyla replied:
Fixed.

dhalperi commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java:
```diff
@@ -796,6 +796,29 @@ public long insertAll(
+              "Unable to patch table description: %s, aborting after %d retries.",
+              tableId, MAX_RPC_RETRIES),
+          Sleeper.DEFAULT,
+          backoff,
```
How do you want this feature to interact with per-window tables created in streaming mode? I think the new table description is not applied in those cases.

ravwojdyla replied:
Good question. Would it make sense to patch description of each table we are streaming to, precisely right after `createdTables.add(tableSpec);`?

dhalperi replied:
I would probably do it here: https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L2724 based on if this worker successfully created the table (return value != null).

dhalperi replied:
Nope, that's not the return type of the createTable API.

Maybe move description into the createTable API proper, where it actually can be embedded in the request we're already sending? https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#description

ravwojdyla replied:
Pushed commit with suggestion above - please let me know what do you think.

ravwojdyla replied:
Sorry missed the comments above. Yes the `createTable` sounds like best idea! Will go ahead and add that.

dhalperi commented:
Looks great to me! Just a few minor nits.

I can also fix those in the merge so feel free to leave as-is; I will fixup and merge tomorrow if you haven't made the fixups.

ravwojdyla commented:
@dhalperi cleaned up a bit:
 * refactored `patchTableDescription` method - now simply accepts `TableReference` instead of project/dataset/table
 * added `@Nullable` where applicable
 * fixed one typo
I am happy to fix anything that is still not right.

ravwojdyla commented:
@dhalperi there definitely was sth going on with GH yesterday. anyway, no problem. fixed your comments.

dhalperi commented:
Thanks Rav!
